Add PWM pulse capture driver for S32K344

Dat-NguyenDuy opened the pull request:
This introduces PWM pulse capture driver over eMIOS peripheral for S32K344

Dat-NguyenDuy committed:
misc: nxp_s32_emios: enable and declare interrupt handler

This enables and declares interrupt handlers for eMIOS,
the handlers defined and implemented at HAL, the driver
takes the name for each id from interrupt-names devicetree

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
tests: pwm_loopback: enable test for mr_canhubk3

Enable PWM loopback testing for mr_canhubk3

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

mmahadevan108 commented on drivers/misc/nxp_s32_emios/nxp_s32_emios.c:
```diff
@@ -87,11 +91,42 @@ static int nxp_s32_emios_init(const struct device *dev)
+		IRQ_CONNECT(DT_IRQ_BY_IDX(node_id, idx, irq),					\
+			    DT_IRQ_BY_IDX(node_id, idx, priority),				\
+			    EMIOS_PWM_INTERRUPT_NAME(DT_STRING_TOKEN_BY_IDX(node_id, prop, idx)),\
+			    DEVICE_DT_GET(node_id),						\
```
Where are these interrupt handlers implemented?

Dat-NguyenDuy replied:
Those are defined in HAL: https://github.com/zephyrproject-rtos/hal_nxp/blob/master/s32/drivers/s32k3/Mcl/src/Emios_Mcl_Ip_Irq.c#L542

manuargue commented on dts/bindings/pwm/nxp,s32-emios-pwm-capture.yaml:
```diff
@@ -0,0 +1,61 @@
+description: |
+  NXP S32 eMIOS PWM pulse capture node for S32 SoCs
+
+compatible: "nxp,s32-emios-pwm-capture"
```
My concern with this approach is that we are introducing two bindings for the same hw block, but each binding is addressing a different functionality, generate and capture, of PWM. I think it would be better to have a single binding and device driver for the EMIOS PWM that is used for both capture and generation. Unless there's something particular to this device (hw) that doesn't allows us to do that.

Dat-NguyenDuy replied:
ok, i've updated to use single shim driver for both

Dat-NguyenDuy committed:
drivers: pwm_nxp_s32_emios: prepare for support pwm capture

This prepares support pwm capture APIs by extended current pwm
shim driver but use a differrence hal component:

- Introduce a Kconfig options that will be set when PWM pulse
  generation API is used, it is also used to select the hal
  component. Guarding current code inside this Kconfig option

- Increase #pwm-cells to 3, flags is supported for PWM capture

- Do not require duty-cycle and polarity be set in dt, PWM
  capture doesn't need it.

- Rename emum value for pwm-mode to keep only key information

- Add preprocessor in case no channel is configured for generate
  PWM output, to avoid warning when build

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
drivers: pwm_nxp_s32_emios: add support for pwm capture

This introduces pwm capture shim driver for NXP S32 EMIOS,
the driver uses SAIC mode that is supported for all channels,
to capture the counter value on each edge for period/pulse
measurement

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

manuargue commented on drivers/pwm/Kconfig.nxp_s32_emios:
```diff
@@ -9,3 +9,15 @@ config PWM_NXP_S32_EMIOS
+	bool "NXP S32 PWM-eMIOS pulse gen driver"
+	default y
+	select NXP_S32_EMIOS
+	select NOCACHE_MEMORY
```
`select NOCACHE_MEMORY if ARCH_HAS_NOCACHE_MEMORY_SUPPORT`

manuargue commented on drivers/pwm/pwm_nxp_s32_emios.c:
```diff
@@ -9,7 +9,10 @@
 #include <zephyr/drivers/clock_control.h>
 #include <Emios_Mcl_Ip.h>
+
+#ifdef CONFIG_PWM_NXP_S32_EMIOS_PULSE_GEN
```
do we really need to build everything conditionally? cannot we assume that PWM generation is always built, as in the rest of the PWM drivers?

Dat-NguyenDuy replied:
Still need conditional because of the glue code. For example no channel is configured for PWM generation such as this https://github.com/zephyrproject-rtos/hal_nxp/pull/277/files#diff-34cb08b7d966e3016d192b073a8796259280fa3cfca2861ae13eb754bf145216R142 will expose to 0 then causes some arrays declared with no element --> warning when build. I've updated to don't need user need to set any Kconfig or the like

manuargue commented on drivers/pwm/pwm_nxp_s32_emios.c:
```diff
@@ -8,12 +8,17 @@
 #include <Emios_Pwm_Ip.h>
 #endif
 
+#ifdef CONFIG_PWM_NXP_S32_EMIOS_PULSE_CAPTURE
```
can we use `CONFIG_PWM_CAPTURE` instead of adding a new option?

Dat-NguyenDuy replied:
I've updated

manuargue commented:
The CI failures are due to https://github.com/zephyrproject-rtos/zephyr/pull/63238, you'll need to rebase once the fix is in

manuargue reviewed (commented):
LGTM overall. @mmahadevan108 do you think we can get the HAL pr merged while we gather more feedback on this?

mmahadevan108 commented:
@Dat-NguyenDuy . Kindly rebase to pull in the HAL SHA

mmahadevan108 commented:
@Dat-NguyenDuy , please also provide on how this PR was tested.

Dat-NguyenDuy commented:
> @Dat-NguyenDuy . Kindly rebase to pull in the HAL SHA

I rebased

Dat-NguyenDuy commented:
> @Dat-NguyenDuy , please also provided details on how this PR was tested.

tested with https://github.com/zephyrproject-rtos/zephyr/pull/62695/files#diff-dea2b57b1ef93f5fb5025176b09c8350e8de1d2e9cb9cb5f9664d211801821d7

manuargue reviewed (dismissed):
All my comments addressed

Dat-NguyenDuy commented:
Rebase to fix conflict. But there would have build fails because of this PR was merged in hal_nxp https://github.com/zephyrproject-rtos/hal_nxp/pull/273. But unfortunately the one in zephyr is still under review https://github.com/zephyrproject-rtos/zephyr/pull/62755

mmahadevan108 commented:
@Dat-NguyenDuy , we have reverted the pinctlrl PR that. Can you please update your west.yml SHA.
